src/concurrent.rs

```diff
@@ -146,104 +30,135 @@ pub(crate) struct FilesData {
-            };
-
-            receivers.push(t);
+        } else if (include.is_empty() || include.is_match(&path))
+            && (exclude.is_empty() || !exclude.is_match(&path))
+            && path.is_file()
+        {
+            self.send_file(path, &sender)?;
         }
 
-        let all_files = match producer.join() {
-            Ok(res) => res,
-            Err(_) => return Err(ConcurrentErrors::Producer("Child thread panicked").into()),
-        };
+        Ok(())
+    }
 
-        // Poison the receiver, now that the producer is finished.
-        for _ in 0..self.num_jobs {
-            if let Err(e) = sender.send(None) {
-                return Err(ConcurrentErrors::Sender(e.to_string()).into());
+    fn consumer(&self, receiver: Receiver<PathBuf>, sender: Sender<Snippets>) -> Result<()> {
+        // Extracts the snippets from the files received from the producer
+        // and sends them to the composer.
+        while let Ok(file) = receiver.recv() {
+            if let Some(snippets) = (self.proc_files)(file.clone(), &self.complexities) {
+                sender
+                    .send(snippets)
+                    .map_err(|e| Concurrent(format!("Sender: {}", e).into()))?;
             }
         }
 
-        for receiver in receivers {
-            if receiver.join().is_err() {
-                return Err(
-                    ConcurrentErrors::Receiver("A thread used to process a file panicked").into(),
-                );
-            }
+        Ok(())
+    }
+
+    fn composer(&self, receiver: Receiver<Snippets>) -> Result<Vec<Snippets>> {
+        let mut snippets_result = Vec::new();
+
+        // Collects the snippets received from the consumer.
+        while let Ok(snippets) = receiver.recv() {
+            snippets_result.push(snippets)
         }
 
-        all_files
+        Ok(snippets_result)
+    }
+
+    /// Runs the producer-consumer-composer approach to process the files
+    /// contained in a directory and in its own subdirectories.
+    ///
+    /// * `files_data` - Information about the files to be included or excluded
+    ///    from a search more the number of paths considered in the search.
+    pub(crate) fn run(self, files_data: FilesData) -> Result<Vec<Snippets>>
+    where
+        Self: Sync,
+    {
+        let (producer_sender, consumer_receiver) = bounded(self.num_jobs);
+        let (consumer_sender, composer_receiver) = bounded(self.num_jobs);
+
+        let result = scope(|scope| {
+            // Producer.
+            scope.spawn(|_| self.producer(producer_sender, files_data));
+
+            // Composer.
+            let composer = scope.spawn(|_| self.composer(composer_receiver));
+
+            // Consumer.
+            (0..self.num_jobs).into_par_iter().try_for_each(|_| {
+                self.consumer(consumer_receiver.clone(), consumer_sender.clone())
+            })?;
+            drop(consumer_sender);
```

Comment:
Is this line necessary? If it is, let's add a comment to specify why.

Reply:
Yes, it is necessary because if the `Sender` is not dropped the channel between **consumers** and **composer** doesn't close and the **composer** will continue to wait to receive data from the **consumers**.

Reply:
Ok, let's add a comment then, thank you!